### PR TITLE
Make 404 pages more useful

### LINKED ISSUE #485: Make 404 pages more useful
**Examples:**

- non-existent branch https://github.com/sindresorhus/refined-github/tree/404-branch
- existent file https://github.com/sindresorhus/refined-github/blob/shorten-url-labels/test/404.js
- deleted file https://github.com/sindresorhus/refined-github/blob/master/extension/content.js

**Possible content:**
- link to repo root
- link to closest existing directory
- link to last commit that had the file ([example](https://github.com/sindresorhus/refined-github/blob/be756e68dd4bf24d53e802e1fec126506979fa80/extension/content.js)) (possible? edit: [yes](https://github.com/sindresorhus/refined-github/issues/485#issuecomment-309264006))

Perhaps this could be a separate extension though.

### PR DESCRIPTION
**Pull request by bfred-it**
Examples:

- 404 file: https://github.com/sindresorhus/refined-github/blob/master/wrong-file

	

- 410 file: https://github.com/sindresorhus/refined-github/blob/master/extension/content.js

	

- 404 branch + file: https://github.com/sindresorhus/refined-github/blob/404/wrong-file

	

- 410 branch + file: https://github.com/sindresorhus/refined-github/blob/404/package.json

	

Closes #485
Replaces, includes and closes #1531
Depends on and includes #1557

### CODE REVIEW

**Comment by bfred-it** [comment]
screen shot 2018-10-03 at 17 39 35

**Comment by DerTieran** [comment]
Nice work, thanks for continuing on my work :+1:
*(Sorry for my PRs going stall, it's been a few busy weeks :sweat_smile:)*

File: source/features/useful-not-found-page.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		<p class="container">
+			See also the file’s {<a href={url}>commit history</a>}
+		</p>
+	);
+}
+
+// If the resource exists in the default branch, link it
+async function addDefaultBranchLink(bar) {
+	const parts = getCleanPathname().split('/');
+	const [,,, branch] = parts;
```
Comment (sindresorhus) [inline source/features/useful-not-found-page.js]: I don't think we should use destructuring when it's more than one lone `,`.

File: source/features/useful-not-found-page.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+			See also the file on the {<a href={url}>default branch</a>}
+		</p>
+	);
+}
+
+export default function () {
+	const parts = parseCurrentURL();
+	const bar = <h2 class="container" />;
+
+	for (let i = 0; i < parts.length; i++) {
```
Comment (sindresorhus) [inline source/features/useful-not-found-page.js]:
```js
for (const [i, part] of parts.entries()) {
```

**Comment by sindresorhus** [comment]
This is incredible useful 🎉 I constantly end up on 404 pages and have to manually modify the URL.

File: source/features/useful-not-found-page.js (hunk 1 above)
Comment (sindresorhus) [reply source/features/useful-not-found-page.js]: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/197

File: source/features/useful-not-found-page.js (hunk 2 above)
Comment (bfred-it) [reply source/features/useful-not-found-page.js]: Thought so, but just a few lines later I’m on a similar loop that can’t easily use for-of (it comes out longer) so I left this one for consistency. What do you think?

Comment (sindresorhus) [reply source/features/useful-not-found-page.js]: Readability > Consistency, and consistency between loop statements doesn't matter much. I still think this should be changed.

**Comment by bfred-it** [comment]
No worries @DerTieran, thanks for starting this!

**Comment by sindresorhus** [comment]
Yay! 🔥

File: source/features/useful-not-found-page.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	const bar = <h2 class="container"/>;
+
+	for (const [i, part] of parts.entries()) {
+		if (i === 2 && part === 'tree') {
+			// `/tree/` is not a real part of the URL
+			continue;
+		}
+		if (i === parts.length - 1) {
+			// The last part of the URL is a known 404
+			bar.append(' / ', getStrikeThrough(part));
```
Comment (yakov116) [inline source/features/useful-not-found-page.js]: Small nitpick if the user is not found then the slash is appended.
Is that intended?

image

**Comment by yakov116** [comment]
Another one is https://github.com/sindresorhus/refined-github/issues/3515

Will ask i you want to see it on the default branch. It will include a link of https://github.com/sindresorhus/refined-github/issues/master Which puts in the search bar `is:open is:issue author:master `

**Comment by bfred-it** [comment]
Excellent, I definitely didn't spend much time thinking about all the possible URLs. Can you open an issue for both? It should probably be limited to `tree` and `blob` URLs.

**Pull request merged**